tools/crm_rule.c

```diff
@@ -0,0 +1,286 @@
+     * to check expressions more complicated than that.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+
+    max = numXpathResults(xpathObj);
+    if (max == 0) {
+        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        rc = 0;
+        goto bail;
+    } else if (max > 1) {
+        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        rc = CRM_EX_ERROR;
+        goto bail;
+    }
+
+    match = getXpathResult(xpathObj, 0);
+
+    /* We should have ensured both of these pass with the xpath query above, but
+     * double checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+    CRM_ASSERT(find_expression_type(match) == time_expr);
+
+    result = eval_date_expression(match, effective_date);
+
+    if (result == date_after_range) {
+        printf("Rule %s is expired\n", rule_id);
+        rc = 1;
+    } else if (result == date_within_range) {
+        printf("Rule %s is still in effect\n", rule_id);
+        rc = 0;
+    } else if (result == date_before_range) {
+        printf("Rule %s has not yet taken effect\n", rule_id);
+        rc = 1;
+    } else {
+        printf("Could not determine whether rule %s is expired\n", rule_id);
+        rc = 1;
+    }
+
+bail:
+    free(xpath);
+    freeXpathObject(xpathObj);
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+    cib_t *cib_conn = NULL;
+
+    int flag = 0;
+    int option_index = 0;
+
+    char *rule_id = NULL;
+    crm_time_t *rule_date = NULL;
+
+    xmlNode *input = NULL;
+    char *input_xml = NULL;
+    char *input_file = NULL;
+    gboolean input_stdin = FALSE;
+
+    int rc = pcmk_ok;
+    crm_exit_t exit_code = CRM_EX_OK;
+
+    crm_log_cli_init("crm_rule");
+    crm_set_options(NULL, "[options]", long_options,
+                    "Tool for querying the state of rules");
+
+    while (flag >= 0) {
+        flag = crm_get_option(argc, argv, &option_index);
+        switch (flag) {
+            case -1:
+                break;
+
+            case 'V':
+                crm_bump_log_level(argc, argv);
+                break;
+
+            case '$':
+            case '?':
+                crm_help(flag, CRM_EX_OK);
+                break;
+
+            case 'c':
+                rule_mode = crm_rule_mode_check;
+                break;
+
+            case 'd':
+                rule_date = crm_time_new(optarg);
+                if (rule_date == NULL) {
+                    exit_code = CRM_EX_DATAERR;
+                    goto bail;
+                }
+
+                break;
+
+            case 'X':
+                input_xml = optarg;
+                break;
+
+            case 'x':
+                input_file = optarg;
+                break;
+
+            case 'p':
+                input_stdin = TRUE;
+                break;
```

Comment:
I have mixed feelings about these options.

We have the CIB_file feature for using an XML file. I'm not sure why some of our commands support a separate command line option for that, possibly the command-line option predated the availability of CIB_file, or it's just duplication. (With the exception of crm_verify, whose purpose is to check whether XML is valid as a CIB, whereas CIB_file presumes that it already is.)

However there's no equivalent for XML from a string or stdin, and those could certainly be useful for this command. I'm thinking maybe: drop -x, and combine the other two into -X where "-X -" indicates stdin (following the convention of many commands that use a dash to indicate stdin/stdout).

Reply:
I patterned these off another program in tools/, thinking there'd be some consistency in command line argument options so I should follow along to preserve it.  I guess there's not really that consistency, though.  Changed.